File: vault_ai_frontend/lib/chunked_aead.dart
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';


const int chunkNonceBytes = 12;
const int chunkTagBytes = 16;
const int chunkIndexAadBytes = 4;

final AesGcm _aesGcm = AesGcm.with256bits();

Uint8List _indexAad(int chunkIndex) {
  if (chunkIndex < 0) {
    throw ArgumentError('chunkIndex must be non-negative');
  }
  if (chunkIndex > 0xFFFFFFFF) {
    throw ArgumentError('chunkIndex exceeds 32-bit range');
  }
  final bd = ByteData(chunkIndexAadBytes);
  bd.setUint32(0, chunkIndex, Endian.big);
  return bd.buffer.asUint8List();
}


Future<Uint8List> encryptChunk({
  required List<int> plaintext,
  required List<int> key,
  required int chunkIndex,
}) async {
  final secretBox = await _aesGcm.encrypt(
    plaintext,
    secretKey: SecretKey(key),
    aad: _indexAad(chunkIndex),
  );
  return _concat([secretBox.nonce, secretBox.cipherText, secretBox.mac.bytes]);
}


Future<Uint8List> encryptChunkWithNonce({
  required List<int> plaintext,
  required List<int> key,
  required int chunkIndex,
  required List<int> nonce,
}) async {
  if (nonce.length != chunkNonceBytes) {
    throw ArgumentError('nonce must be exactly $chunkNonceBytes bytes');
  }
  final secretBox = await _aesGcm.encrypt(
    plaintext,
    secretKey: SecretKey(key),
    nonce: nonce,
    aad: _indexAad(chunkIndex),
  );
  return _concat([secretBox.nonce, secretBox.cipherText, secretBox.mac.bytes]);
}


Future<Uint8List> decryptChunk({
  required List<int> frame,
  required List<int> key,
  required int chunkIndex,
}) async {
  if (frame.length < chunkNonceBytes + chunkTagBytes) {
    throw ArgumentError('Chunk frame too short');
  }
  final nonce = frame.sublist(0, chunkNonceBytes);
  final mac = frame.sublist(frame.length - chunkTagBytes);
  final ct = frame.sublist(chunkNonceBytes, frame.length - chunkTagBytes);
  final secretBox = SecretBox(ct, nonce: nonce, mac: Mac(mac));
  final plaintext = await _aesGcm.decrypt(
    secretBox,
    secretKey: SecretKey(key),
    aad: _indexAad(chunkIndex),
  );
  return Uint8List.fromList(plaintext);
}

Uint8List _concat(List<List<int>> parts) {
  final total = parts.fold<int>(0, (acc, p) => acc + p.length);
  final out = Uint8List(total);
  var offset = 0;
  for (final p in parts) {
    out.setRange(offset, offset + p.length, p);
    offset += p.length;
  }
  return out;
}

File: vault_ai_backend/chunked_aead.py


from __future__ import annotations

import os
import struct

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


CHUNK_NONCE_BYTES = 12
CHUNK_TAG_BYTES = 16
CHUNK_INDEX_AAD_BYTES = 4


def _index_aad(chunk_index: int) -> bytes:
    if chunk_index < 0:
        raise ValueError("chunk_index must be non-negative")
    if chunk_index > 0xFFFFFFFF:
        raise ValueError("chunk_index exceeds 32-bit range")
    return struct.pack(">I", chunk_index)


def encrypt_chunk(plaintext: bytes, key: bytes, chunk_index: int) -> bytes:


    aesgcm = AESGCM(key)
    nonce = os.urandom(CHUNK_NONCE_BYTES)
    ct_and_tag = aesgcm.encrypt(nonce, plaintext, _index_aad(chunk_index))
    return nonce + ct_and_tag


def decrypt_chunk(frame: bytes, key: bytes, chunk_index: int) -> bytes:


    if len(frame) < CHUNK_NONCE_BYTES + CHUNK_TAG_BYTES:
        raise ValueError("Chunk frame too short")
    nonce = frame[:CHUNK_NONCE_BYTES]
    ct_and_tag = frame[CHUNK_NONCE_BYTES:]
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(nonce, ct_and_tag, _index_aad(chunk_index))


def _encrypt_chunk_with_nonce(
    plaintext: bytes, key: bytes, chunk_index: int, nonce: bytes
) -> bytes:
    if len(nonce) != CHUNK_NONCE_BYTES:
        raise ValueError("nonce must be exactly 12 bytes")
    aesgcm = AESGCM(key)
    ct_and_tag = aesgcm.encrypt(nonce, plaintext, _index_aad(chunk_index))
    return nonce + ct_and_tag
